python/tests/test_combinatorics.py

```diff
@@ -0,0 +1,668 @@
+        Given a rank and the number of leaves, produce the corresponding tree.
+
+        WARNING: This does not produce properly labeled trees.
+        """
+        assert rank < comb.S(n)
+
+        if n == 1:
+            assert rank == 0
+            return Leaf(0)
+
+        k, l_rank, r_rank = comb.subtree_shape_ranks(rank, n)
+        left = Tree.shape_unrank(l_rank, k)
+        right = Tree.shape_unrank(r_rank, n - k)
+        return Tree(left, right)
+
+    def label_unrank(self, rank, labels=None):
+        """
+        Given the rank of a labeling, assign the appropriate labels
+        to the leaves of this tree.
+        """
+        if labels is None:
+            labels = list(range(self.num_leaves()))
+
+        left_L = L_tree(self.left)
+        right_L = L_tree(self.right)
+
+        comb_rank = rank // (left_L * right_L)
+        rank_given_comb = rank % (left_L * right_L)
+
+        left_label_rank = rank_given_comb // right_L
+        right_label_rank = rank_given_comb % right_L
+        left_labels = comb.Labeling.unrank(comb_rank, labels, self.left.num_leaves())
+        right_labels = [lab for lab in labels if lab not in set(left_labels)]
+
+        self.left.label_unrank(left_label_rank, left_labels)
+        self.right.label_unrank(right_label_rank, right_labels)
+
+    ########################################################################
+    # Helpers
+    ########################################################################
+
+    def is_symmetrical(self):
+        return (
+            self.left.num_leaves() == self.right.num_leaves()
+            and self.left.shape_rank() == self.right.shape_rank()
+        )
+
+    def is_canonical(self):
+        return (
+            self.left.num_leaves() >= self.right.num_leaves()
+            and self.left.is_canonical()
+            and self.right.is_canonical()
+        )
+
+    def canonical(self):
+        left_canon = self.left.canonical()
+        right_canon = self.right.canonical()
+        if left_canon.num_leaves() > right_canon.num_leaves():
+            return Tree(left_canon, right_canon)
+        else:
+            return Tree(right_canon, left_canon)
+
+    def leaves(self):
+        return heapq.merge(self.left.leaves(), self.right.leaves())
+
+    def shape_equal(self, other):
+        return (
+            isinstance(other, Tree)
+            and self.left.shape_equal(other.left)
+            and self.right.shape_equal(other.right)
+        )
+
+    def equal(self, other):
+        return (
+            isinstance(other, Tree)
+            and self.left.equal(other.left)
+            and self.right.equal(other.right)
+        )
+
+    @staticmethod
+    def from_tsk_tree(tree):
+        def subtree(node_id):
+            if tree.is_leaf(node_id):
+                return Leaf(node_id)
+            left = tree.left_child(node_id)
+            right = tree.right_child(node_id)
+            if tree.num_samples(left) < tree.num_samples(right):
```

Comment:
num_samples and num_leaves are not quite the same thing as we can have internal samples. I think for this work we can assume (or should enforce) that leaves === samples.